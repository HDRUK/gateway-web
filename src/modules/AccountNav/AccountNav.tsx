import Menu from "@mui/material/Menu";
import MenuItem from "@mui/material/MenuItem";
import Link from "@/components/Link";
import Button from "@/components/Button";
import ChevronLeftIcon from "@mui/icons-material/ChevronLeft";
import useAuth from "@/hooks/useAuth";

import { useEffect } from "react";

interface AccountNavProps {
    onCloseMenu: () => void;
    onLogout: () => void;
    anchorElement: null | HTMLElement;
}

const AccountNav = ({
    anchorElement,
    onCloseMenu,
    onLogout,
}: AccountNavProps) => {
    let links = [
        { id: -99, label: "Profile", href: "/account/profile" },
    ];

    const { user } = useAuth();

    const handleCloseUserMenu = () => {
        if (typeof onCloseMenu === "function") {
            onCloseMenu();
        }
    };

    const handleLogout = () => {
        if (typeof onLogout === "function") {
            onLogout();
        }
    };

    useEffect(() => {
        if (!user) {
            return;
        }
    }, [user]);

    user.teams.map(team => {
        let t = { id: team.id, label: team.name, href: `/account/team/${team.id}` };

        if (!links.includes(t)) {
            links.push(t);
        }
    });

    return (
        <Menu
            PaperProps={{
                elevation: 0,
                sx: {
                    overflow: "visible",
                    mt: 1.5,
                    "& .MuiAvatar-root": {
                        width: 32,
                        height: 32,
                        ml: -0.5,
                        mr: 1,
                    },
                    "&:before": {
                        content: '""',
                        display: "block",
                        position: "absolute",
                        top: 0,
                        right: 14,
                        width: 10,
                        height: 10,
                        bgcolor: "background.paper",
                        transform: "translateY(-50%) rotate(45deg)",
                        zIndex: 0,
                    },
                },
            }}
            id="account-nav"
            anchorEl={anchorElement}
            transformOrigin={{ horizontal: "right", vertical: "top" }}
            anchorOrigin={{ horizontal: "right", vertical: "bottom" }}
            onClose={handleCloseUserMenu}
            open={Boolean(anchorElement)}>
            {links.map(link => (
                <Link key={link.label} underline="hover" href={link.href}>
                    <MenuItem
                        sx={{ width: 220 }}
                        key={link.label}
                        LinkComponent={Link}
                        onClick={handleCloseUserMenu}>
                        <ChevronLeftIcon />
                        {link.label}
                    </MenuItem>
                </Link>
            ))}
            <Button variant="link">
                <MenuItem sx={{ width: 220 }} onClick={handleLogout}>
                    Logout
                </MenuItem>
            </Button>
        </Menu>
    );
};

export default AccountNav;
