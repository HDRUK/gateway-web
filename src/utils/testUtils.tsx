import React, { ReactNode } from "react";
import {
    render,
    renderHook,
    RenderHookOptions,
    RenderOptions,
    RenderResult,
    RenderHookResult,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { CacheProvider, ThemeProvider } from "@emotion/react";
import ActionBarProvider from "@/providers/ActionBarProvider";
import DialogProvider from "@/providers/DialogProvider";
import theme from "@/config/theme";
import createEmotionCache from "@/config/createEmotionCache";

const clientSideEmotionCache = createEmotionCache();

const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
        <SWRConfig
            value={{
                provider: () => new Map(),
            }}>
            <CacheProvider value={clientSideEmotionCache}>
                <ThemeProvider theme={theme}>
                    <ActionBarProvider>
                        <DialogProvider>{children}</DialogProvider>
                    </ActionBarProvider>
                </ThemeProvider>
            </CacheProvider>
        </SWRConfig>
    );
};

interface OptionProps
    extends RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRender = (
    ui: React.ReactElement<
        unknown,
        string | React.JSXElementConstructor<unknown>
    >,
    options?: OptionProps
): RenderResult => {
    const { wrapperProps, ...rest } = options || {};

    return render(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

interface OptionHookProps
    extends RenderHookOptions<
        unknown,
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRenderHook = (
    ui: (initialProps: unknown) => unknown,
    options?: OptionHookProps
): RenderHookResult<unknown, unknown> => {
    const { wrapperProps, ...rest } = options || {};
    return renderHook(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

export * from "@testing-library/react";

export { customRenderHook as renderHook };
export { customRender as render };
