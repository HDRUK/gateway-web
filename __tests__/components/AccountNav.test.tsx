import React, { useState } from "react";
import AccountNav from "@/modules/AccountNav";
import { fireEvent, render, screen, waitFor } from "../testUtils";

const DemoComponent = (props: {
    onCloseMenu?: () => void;
    onLogout?: () => void;
}) => {
    const [anchorElement, setAnchorElement] = useState<null | HTMLElement>(
        null
    );
    const handleLogout = () => null;

    return (
        <>
            <input
                type="button"
                value="open nav"
                onClick={e => setAnchorElement(e.currentTarget)}
            />
            <AccountNav
                anchorElement={anchorElement}
                onCloseMenu={() => setAnchorElement(null)}
                onLogout={handleLogout}
                {...props}
            />
        </>
    );
};

DemoComponent.defaultProps = {
    onLogout: () => null,
    onCloseMenu: () => null,
};

describe("AccountNav", () => {
    it("should render component", async () => {
        const wrapper = render(<DemoComponent />);

        expect(wrapper.container).toMatchSnapshot();
    });

    it("should call logout", async () => {
        const onLogout = jest.fn();

        render(<DemoComponent onLogout={onLogout} />);

        fireEvent.click(screen.getByText("open nav"));

        await waitFor(() => {
            expect(screen.getByText("Logout")).toBeInTheDocument();
        });

        fireEvent.click(screen.getByText("Logout"));

        expect(onLogout).toHaveBeenCalled();
    });

    it("should call onCloseMenu", async () => {
        const onCloseMenu = jest.fn();

        render(<DemoComponent onCloseMenu={onCloseMenu} />);

        fireEvent.click(screen.getByText("open nav"));

        await waitFor(() => {
            expect(screen.getByText("Profile")).toBeInTheDocument();
        });

        fireEvent.click(screen.getByText("Profile"));

        expect(onCloseMenu).toHaveBeenCalled();
    });
});
