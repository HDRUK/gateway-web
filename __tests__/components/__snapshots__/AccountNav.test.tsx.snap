// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountNav should render component 1`] = `
<div>
  <input
    type="button"
    value="open nav"
  />
</div>
`;
